R/ModuleEnv.R

```diff
@@ -0,0 +1,72 @@
+
+
+.handleInjectVar <- function (x) {
+  hVar <- function(x) {
+    if (is.character(x)) {
+      return(sprintf('"%s"', x))
+    } else {
+      return(sprintf('%s', x))
+    }
+  }
+
+  if (length(x) == 1) {
+    return(hVar(x))
+  } else if (is.vector(x)) {
+    innerVars <- hVar(x)
+    return(paste0("c(", paste(innerVars, collapse = ", "), ")"))
+  } else {
+    stop("cannot handle complex data structures in variable injection")
+  }
+}
+
+#' Load module execution space inside and renv
+#' inspired by targets::tar_script but allowing custom variable execution
+#'
+#' Designed to allow more human readable code that is executed inside a module as well as simple variable substituion
+#' for injecting constants (e.g. simple parameters or file paths used inside and outside of modules)
+#'
+#' This pattern also allows dependency injection which could be used if you don't want to use and renv and (instead)
+#' would like to use docker images or just execution in the base environment for testing/debugging
+#'
+#' @param code              code block to execute
+#' @param moduleFolder      Instantiated Strategus module folder
+#' @param injectVars         list of var names list(name=value) to replace (e.g. replace list(foo = "some string") will
+#'                          find the pattern foo and replace it with the string some string - be careful!
+#' @param tempScriptFile    tempFile to write script to (ret
+#' @param job               run as rstudio job
+#' @param processName       String name for process
+#' @returns NULL invisibly
+withModuleRenv <- function(code,
+                           moduleFolder,
+                           injectVars = list(),
+                           tempScriptFile = tempfile(fileext = ".R"),
+                           useLocalStrategusLibrary = TRUE,
+                           job = FALSE,
+                           processName = paste(moduleFolder, "_renv_run")) {
+  # convert human readable code to a string for writing
+  script <- as.character(substitute(code))[-1]
+  # Insert variables
+  for (name in names(injectVars)) {
+    rep <- .handleInjectVar(injectVars[[name]])
+    script <- gsub(name, rep, script)
+  }
+
+  # Enforce attachment of Strategus from calling process - note one inside the renv
+  if (useLocalStrategusLibrary) {
+    libPath <- file.path(find.package("Strategus"), "../")
+    script <- c(sprintf("library(Strategus, lib.loc = '%s')", libPath),
```

Comment:
@anthonysena this is essentially how the library call to load Strategus is fixed on unix environments.

This `libPath` usage means that the call to library(Strategus) now knows to use a specific location on disk.
This will not work well with calls for external libraries used by strategus though. That could create some strange conflicts or, if they aren't in the modules renv, they wouldn't be finable.

Reply:
Thanks @azimov - making a note to myself that we'll likely need to add `keyring` to the list of libraries here.